File: 501.find-mode-in-binary-search-tree.java
/*
 * @lc app=leetcode id=501 lang=java
 *
 * [501] Find Mode in Binary Search Tree
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int[] findMode(TreeNode root) {
        int maxTimes = -1;
        int times = -1;
        List<Integer> ret = new ArrayList<>();
        TreeNode prev = null;
        Stack<TreeNode> stack = new Stack<>();
        stack.add(root);
        while (!stack.empty()) {
            TreeNode node = stack.pop();
            if (node != null) {
                if (node.right != null) stack.add(node.right);
                stack.add(node); stack.add(null);
                if (node.left != null) stack.add(node.left);
            } else {
                node = stack.pop();
                if (prev != null) {
                    if (node.val == prev.val) times += 1;
                    else times = 1;
                    if (times == maxTimes) ret.add(node.val);
                    if (times > maxTimes) {
                        ret.clear();
                        ret.add(node.val);
                        maxTimes = times;
                    }
                } else {
                    ret.add(node.val); times = 1; maxTimes = 1;
                }
                prev = node;
            }
        }
        return ret.stream().mapToInt(x -> x).toArray();
    }
}
// @lc code=end

File: 501.find-mode-in-binary-search-tree.cpp
/*
 * @lc app=leetcode id=501 lang=cpp
 *
 * [501] Find Mode in Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:

    vector<int> findMode(TreeNode* root) {
        if (root == nullptr) return {};
        stack<TreeNode *> st; st.push(root);

        TreeNode *prev = nullptr;
        unsigned times = 1, max_times = 1;
        vector<int> ret;


        while (!st.empty()) {
            auto node = st.top(); st.pop();
            if (node != nullptr) {
                if (node->right != nullptr) st.push(node->right);
                st.push(node); st.push(nullptr);
                if (node->left != nullptr) st.push(node->left);
            } else {
                node = st.top(); st.pop();
                if (prev == nullptr) {
                    prev = node; ret.push_back(node->val); continue;
                }
                (node->val == prev->val) ? times += 1 : times = 1;
                prev = node;
                if (times > max_times) {
                    ret.clear(); ret.push_back(node->val); max_times = times;
                } else if (times == max_times) {
                    ret.push_back(node->val);
                }
            }
        }
        return ret;
    }
};
// @lc code=end
